zcash_extensions/src/transparent/demo.rs

```diff
@@ -0,0 +1,734 @@
+                        tze_out.precondition.mode,
+                        &tze_out.precondition.payload,
+                    ) {
+                        Ok(Precondition::Close(p_close)) => {
+                            // Finally, check the precondition:
+                            // precondition_open = BLAKE2b_256(witness_open || precondition_close)
+                            let mut h = Params::new().hash_length(32).to_state();
+                            h.update(&w_open.0);
+                            h.update(&p_close.0);
+                            let hash = h.finalize();
+                            if hash.as_bytes() == p_open.0 {
+                                Ok(())
+                            } else {
+                                Err(Error::HashMismatch)
+                            }
+                        }
+                        Ok(Precondition::Open(_)) => Err(Error::ExpectedClose),
+                        Err(e) => Err(e),
+                    },
+                    _ => Err(Error::InvalidOutputQty(outputs.len())),
+                }
+            }
+            (Precondition::Close(p), Witness::Close(w)) => {
+                // In CLOSE mode, we only require that the precondition is satisfied:
+                // precondition_close = BLAKE2b_256(witness_close)
+                let hash = Params::new().hash_length(32).hash(&w.0);
+                if hash.as_bytes() == p.0 {
+                    Ok(())
+                } else {
+                    Err(Error::HashMismatch)
+                }
+            }
+            _ => Err(Error::ModeMismatch),
+        }
+    }
+}
+
+fn builder_hashes(preimage_1: &[u8; 32], preimage_2: &[u8; 32]) -> ([u8; 32], [u8; 32]) {
+    let hash_2 = {
+        let mut hash = [0; 32];
+        hash.copy_from_slice(Params::new().hash_length(32).hash(preimage_2).as_bytes());
+        hash
+    };
+
+    let hash_1 = {
+        let mut hash = [0; 32];
+        hash.copy_from_slice(
+            Params::new()
+                .hash_length(32)
+                .to_state()
+                .update(preimage_1)
+                .update(&hash_2)
+                .finalize()
+                .as_bytes(),
+        );
+        hash
+    };
+
+    (hash_1, hash_2)
+}
+
+pub struct DemoBuilder<B> {
+    pub txn_builder: B,
+    pub extension_id: u32,
+}
+
+#[derive(Debug)]
+pub enum DemoBuildError<E> {
+    BaseBuilderError(E),
+    ExpectedOpen,
+    ExpectedClose,
+    PrevoutParseFailure(Error),
+    TransferMismatch {
+        expected: [u8; 32],
+        actual: [u8; 32],
+    },
+    CloseMismatch {
+        expected: [u8; 32],
+        actual: [u8; 32],
+    },
+}
+
+impl<'a, B: ExtensionTxBuilder<'a>> DemoBuilder<&mut B> {
+    pub fn demo_open(
+        &mut self,
+        value: Amount,
+        preimage_1: [u8; 32],
+        preimage_2: [u8; 32],
+    ) -> Result<(), DemoBuildError<B::BuildError>> {
+        let (hash_1, _) = builder_hashes(&preimage_1, &preimage_2);
```

Comment:
Similarly here, you'd have `hash_1` but neither of the preimages. For some TZEs, the equivalent to `hash_1` would be something akin to a transparent P2SH address.